### PR TITLE
End test

### LINKED ISSUE #23: Add E2E tests for critical user workflows
## Description
With the E2E framework set up (#22 ), write E2E tests covering the
critical user journeys.

**Parent user stories:**
- #20 

**Grandparent epic:** #19 

## Context
Per the testing guide, E2E coverage targets the workflows a real user cares
about most. Unit and integration tests handle exhaustive coverage; E2E
proves the pieces work together from the browser to the database.

## Critical workflows to cover
1. **Student books a consultation**
   Log in as student → select a lecturer → pick an available slot → submit
   → see confirmation → consultation appears on dashboard.
2. **Student joins an existing consultation**
   Log in → dashboard → join an upcoming consultation that is not full →
   confirm the student's name appears in the attendee list.
3. **Lecturer sets availability**
   Log in as lecturer → set work-week availability + max students +
   duration + max bookings per day → confirm settings persist after reload.
4. **Booking validation rejects invalid requests**
   Attempt to book outside availability → rejection.
   Attempt to join a full consultation → join disabled or rejected.
5. **Cancellation flows both ways**
   Organiser cancels → consultation disappears from both dashboards.
   Lecturer cancels → consultation disappears from student dashboard.

## Tasks
- [ ] Audit the UI for missing `data-testid` attributes on elements the
      tests need. Add any that are missing.
- [ ] Set up per-test database seeding so each test runs against known
      data and leaves no residue.
- [ ] Write one test per workflow above.
- [ ] Ensure all E2E tests run in CI on every PR.
- [ ] Verify independence: run the suite in random order locally; all
      tests must still pass.

## Acceptance Tests
- [ ] All 5 workflow tests pass on a fresh `git clone` after `npm install`
      and `npm run test:e2e`.
- [ ] CI pipeline runs all E2E tests on every PR to main and blocks merge
      if any fail.
- [ ] Running the E2E suite twice in a row both pass (no flakiness from
      leftover state).
- [ ] Each test passes when run in isolation via the framework's filter.

## Notes
Expect some refactoring to add `data-testid` hooks. If a workflow turns
out to be too brittle to test reliably, document the limitation in the
retrospective and fall back to an integration test at the route level
rather than merging a flaky E2E test.

### LINKED ISSUE #98: Cancellation of meetings
## User story
As a student or lecturer, I want to be able to cancel or leave a consultation 
I am part of, so that the slot is freed up for others and dashboards stay accurate.

## Context
The consultation detail page (#48) renders Cancel and Leave buttons but routes 
them to placeholder handlers that return "not yet implemented". This was flagged 
in the Sprint 3 PR review. Cancel is for the organiser; Leave is for a non-organiser 
attendee. Both must update the consultation status and remove the attendee row 
where appropriate. Cancellation by the organiser should mark status as `Cancelled` 
and remove all attendees. Leaving by a non-organiser should remove only that 
student from `consultation_attendees`.

## Acceptance tests
- [ ] Organiser clicks Cancel → status becomes `Cancelled`, attendees cleared, 
      consultation disappears from all dashboards
- [ ] Non-organiser attendee clicks Leave → their row removed from 
      `consultation_attendees`, consultation remains for others
- [ ] After cancellation, the time slot is freed and appears as bookable again 
      on the student dashboard calendar
- [ ] A student who has left a consultation can re-book or join another slot 
      in the same window
- [ ] Cancel and Leave are not visible to students who are not the organiser 
      or an attendee respectively

**Labels:** `user-story`  
**Story points:** 5  
**Sprint:** 4  
**Parent:** #1 (Book and manage consultations)

### PR DESCRIPTION
**Pull request by Aditya-Raghunandan**
## Summary

- Implements cancel and leave for story #98: organiser cancel sets status `Cancelled` and clears all attendees; non-organiser attendee leave removes only their row from `consultation_attendees`
- Adds E2E test suite for story #23: booking validation, lecturer availability add/delete, organiser cancel, non-organiser leave — all passing

## Changes

**`src/controllers/consultation-detail-controller.js`**
- `cancelConsultation`: now deletes all rows from `consultation_attendees` after marking status `Cancelled`
- `leaveConsultationTodo` → `leaveConsultation`: fully implemented — guards that user is an attendee but not organiser, then removes their row

**`src/routes/consultation-routes.js`**
- Updated import/route to use renamed `leaveConsultation`

**`src/views/consultation-detail.html`, `src/views/availability.html`**
- Added `data-testid` to cancel, join, leave buttons and availability table/save-slot button

**`tests/e2e/`** — 3 files, 11 tests (1 skipped: join via GET link — pending PR)
- `cancellation.spec.js`: organiser cancel, non-organiser leave, button visibility enforcement
- `lecturer-availability.spec.js`: add slot → appears in table, delete slot → removed
- `booking-flow.spec.js`: past-date validation redirect, missing-field browser validation, existing tests

## Acceptance criteria (story #98)

- [x] Organiser clicks Cancel → status `Cancelled`, attendees cleared, disappears from dashboards
- [x] Non-organiser attendee clicks Leave → their row removed, consultation remains for others
- [x] After cancellation slot is freed (status filter excludes `Cancelled` from booking calendar)
- [x] Student who left can re-book the same window
- [x] Cancel not visible to non-organisers; Leave not visible to organisers or non-attendees

Closes #23, closes #98

### CODE REVIEW

File: src/controllers/consultation-detail-controller.js
Comment (ThandiAlgorithm) [inline src/controllers/consultation-detail-controller.js:1]: Great work on this file. The cancel and leave behaviours are now much clearer and more complete. I like that organiser cancellation and attendee leaving are treated as separate flows, because they should not affect the database in the same way.

1. Cancelling now updates the consultation status to `Cancelled`.
2. Cancelling also clears all attendees, which matches the story requirement.
3. `leaveConsultation` replaces the old placeholder properly.
4. The leave flow has good guard checks for missing consultations, organisers, and non-attendees.

File: src/routes/consultation-routes.js [hunk 1]
```diff
@@ -14,6 +14,6 @@ router.post('/consultations/new', createBooking);
 router.get('/consultations/:constId', showConsultationDetail);
 router.post('/consultations/:constId/join', joinConsultation);
 router.post('/consultations/:constId/cancel', cancelConsultation);
-router.post('/consultations/:constId/leave', leaveConsultationTodo);
+router.post('/consultations/:constId/leave', leaveConsultation);
 
 module.exports = router;
```
Comment (ThandiAlgorithm) [inline src/routes/consultation-routes.js:19]: Good update. This route now points to the real leave implementation and keeps the action as a POST request, which is the correct method for changing attendance state.

File: src/views/availability.html [hunk 2]
```diff
@@ -106,7 +106,7 @@ <h5 class="fw-bold mb-3">Add New Slot</h5>
                   name="max_number_of_students" min="1" max="10" value="1" required>
               </div>
 
-              <button type="submit" class="btn btn-primary w-100">Save Slot</button>
+              <button type="submit" class="btn btn-primary w-100" data-testid="save-slot-btn">Save Slot</button>
```
Comment (ThandiAlgorithm) [inline src/views/availability.html:109]: data-testid="save-slot-btn" gives the add-slot test a stable selector

File: src/views/availability.html [hunk 3]
```diff
@@ -121,7 +121,7 @@ <h5 class="fw-bold mb-3">Your Availability Slots</h5>
               </div>
             <% } else { %>
               <div class="table-responsive">
-                <table class="table table-hover align-middle mb-0">
+                <table class="table table-hover align-middle mb-0" data-testid="availability-table">
```
Comment (ThandiAlgorithm) [inline src/views/availability.html:124]: data-testid="availability-table" gives the test a stable place to verify saved slots

File: src/views/consultation-detail.html
Comment (ThandiAlgorithm) [inline src/views/consultation-detail.html:1]: Great small change with a big testing benefit. Adding test IDs to the action buttons makes the cancellation, leave, and join flows much easier to test from the browser.

File: tests/e2e/booking-flow.spec.js
Comment (ThandiAlgorithm) [inline tests/e2e/booking-flow.spec.js:1]: Strong additions here. These tests help protect important booking edge cases and give confidence that invalid bookings are handled correctly through the browser. A future improvement could also assert that no consultation row was created in the database.

File: tests/e2e/cancellation.spec.js
Comment (ThandiAlgorithm) [inline tests/e2e/cancellation.spec.js:1]: This is a very valuable test file. It covers the cancellation and leave story through the real browser flow, which gives much stronger confidence than only testing the controller.

File: tests/e2e/lecturer-availability.spec.js
Comment (ThandiAlgorithm) [inline tests/e2e/lecturer-availability.spec.js:1]: Great practical E2E coverage. This file proves that the lecturer availability page works for both adding and deleting slots, which is one of the core lecturer workflows.

**Review by ThandiAlgorithm — COMMENTED** [review]
Great PR. This completes an important cancellation and leave workflow while also improving E2E test coverage across student and lecturer flows. The only improvement I would strongly consider is wrapping the cancel status update and attendee deletion in a transaction so the database cannot end up partially updated.

File: src/routes/consultation-routes.js (hunk 1 above)
Comment (Aditya-Raghunandan) [reply src/routes/consultation-routes.js:19]: Nice

File: tests/e2e/lecturer-availability.spec.js
Comment (Aditya-Raghunandan) [reply tests/e2e/lecturer-availability.spec.js:1]: 👍
